lib/ansible/plugins/inventory/__init__.py

```diff
@@ -292,21 +293,28 @@ def _add_host_to_keyed_groups(self, keys, variables, host, strict=False):
             for keyed in keys:
                 if keyed and isinstance(keyed, dict):
                     prefix = keyed.get('prefix', '')
+                    if prefix:
+                        prefix = prefix + "_"
                     key = keyed.get('key')
                     if key is not None:
                         try:
-                            groups = to_safe_group_name('%s_%s' % (prefix, self._compose(key, variables)))
+                            groups = self._compose(key, variables)
                         except Exception as e:
                             if strict:
                                 raise AnsibleOptionsError("Could not generate group on %s: %s" % (key, to_native(e)))
                             continue
-                        if isinstance(groups, string_types):
-                            groups = [groups]
                         if isinstance(groups, list):
                             for group_name in groups:
-                                if group_name not in self.inventory.groups:
-                                    self.inventory.add_group(group_name)
-                                self.inventory.add_child(group_name, host)
+                                safe_group_name = to_safe_group_name('%s%s' % (prefix, group_name))
+                                if safe_group_name not in self.inventory.groups:
+                                    self.inventory.add_group(safe_group_name)
+                                self.inventory.add_child(safe_group_name, host)
+                        elif isinstance(groups, dict):
```

Comment:
This no longer works consistently with values. It works with values that are in {key: value} format, but not [{key: value}]. If a user wants to create specific groups from ProductCodes/BlockDeviceMappings/ElasticGpuAssociations/NetworkInterfaces/SecurityGroups/Tags (i.e. anything that is a list of dicts), the value is ignored and all hosts with the key are added to the group.
reproducer:
```
plugin: aws_ec2
regions:
  - us-east-1
keyed_groups:
  - key: 'SecurityGroups'
    value: {'group_id': 'sg-xxxxxxxx'}
```
returns all instances in us-east-1 in the SecurityGroups group, regardless of the group id.

Reply:
Did this work before this change?

Reply:
```
keyed_groups:
  - key: 'security_groups|json_query("[].group_id")'
    prefix: 'security_groups'
```

works as expected. I'm not really sure how `value` actually provides filtering here, or why you'd want that vs. using `filters` section.

Reply:
This is broken on devel, you can disregard the comment. Not related to your PR.